### PR TITLE
Add auto fixture to use tmp db per session

### PR DESCRIPTION
**Pull request by davzucky**
## Summary
This is a new fixture that creates a tmp per pytest run.  I had on going dev with another version of Orion and all the pytest are failing because of that. That fixture allow to have a tmp db of orion everytime

I expect we could extract this fixture to a package pytest-prefect and can be shared with another prefect project in the future

## Relevant Issue(s)
This is an enhancement to the current pytest capability

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -10,3 +10,18 @@ def aws_credentials():
...
+@pytest.fixture(scope="session")
+def temp_db_path(tmpdir_factory):
+    tmp_db_path = tmpdir_factory.mktemp("db")
+    db_file_path = tmp_db_path.join("orion.db")
+    yield str(db_file_path)
+
+
+@pytest.fixture(autouse=True)
+def init_prefect_db(monkeypatch, temp_db_path):
+    monkeypatch.setenv("PREFECT_ORION_DATABASE_CONNECTION_URL",f"sqlite+aiosqlite://{temp_db_path}")
```
Comment (ahuang11) [inline tests/conftest.py:24]:
```suggestion
    monkeypatch.setenv("PREFECT_ORION_DATABASE_CONNECTION_URL", f"sqlite+aiosqlite://{temp_db_path}")
```

File: tests/conftest.py [hunk 2]
```diff
@@ -10,3 +10,18 @@ def aws_credentials():
...
+    yield
```
Comment (ahuang11) [inline tests/conftest.py:25]: Wondering if we still need a `yield` here?

**Comment by ahuang11** [comment]
Thanks for another contribution!

I think we need to fix the CI tools which I can tackle tomorrow.
https://github.com/PrefectHQ/prefect/issues/5609

Also want to give Alex some time to review.

File: tests/conftest.py (hunk 1 above)
Comment (davzucky) [reply tests/conftest.py:24]: Updated. Running black fixed that. Interesting that pre-commit didn't catch it.

File: tests/conftest.py (hunk 2 above)
Comment (davzucky) [reply tests/conftest.py:25]: Removed the yield. It was from the first implementation in which I was cleaning the db folder.

**Comment by davzucky** [comment]
@ahuang11 Any thought about moving that fixture to its own python package to share with other collections?

**Review by zangell44 — COMMENTED** [review]
It hasn't been released yet but we just introduced a utility in `prefect` for this. Usage would look like

```python
import pytest
from prefect.utilities.testing import prefect_test_harness

@pytest.fixture(autouse=True)
def prefect_db():
     with prefect_test_harness():
         yield
```

There's some additional considerations under the hood handled by the test harness, I would recommend using that going forward

**Comment by davzucky** [comment]
> 

Thank you @zangell44 for the information. Great to know you have a helper function that will come with it. Could you share what hedge case may be missed by my solution?
I was really only focused on isolating the orion db at the moment, and I would like to understand better what other considerations I should have in the future.

**Comment by zangell44** [comment]
Yep happy to share @davzucky!
The main two considerations during tests are
- Ensuring Prefect does not talk to an API endpoint (especially the Cloud endpoint)
- Ensuring our database interface is properly configured (really not a user facing consideration but quite important for Prefect's internals supporting Postgres/SQLite/additional customization)


The helper function looks like this (omitting imports as at least one of them is also not released yet)

```python
@contextmanager
def prefect_test_harness():
    """
    Temporarily run flows against a local SQLite database for testing.

    Example:
        >>> from prefect import flow
        >>> @flow
        >>> def my_flow():
        >>>     return 'Done!'
        >>> with prefect_test_harness():
        >>>     assert my_flow() == 'Done!' # run against temporary db
    """
    # create temp directory for the testing database
    with TemporaryDirectory() as temp_dir:
        with ExitStack() as stack:
            # temporarily override any database interface components
            stack.enter_context(temporary_database_interface())
            # set the connection url to our temp database
            # make sure PREFECT_API_URL is not set, otherwise
            # actual API requests will be made
            stack.enter_context(
                temporary_settings(
                    PREFECT_API_URL=None,
                    PREFECT_ORION_DATABASE_CONNECTION_URL=f"sqlite+aiosqlite:////{temp_dir}/orion.db",
                )
            )
            yield
```

Most important is making sure `PREFECT_API_URL` is not set during tests. If it is set, any requests made by the `OrionClient` will hit the api endpoint even if you've configured a temporary database.

**Comment by davzucky** [comment]
Thank you for sharing this information. This is really similar to the setup we did with prefect 1.0. Will add the ```PREFECT_API_URL``` in my current change and will update the fixture after you release the code.

**Comment by davzucky** [comment]
One comment about your code, I currently set up the db fixture at the scope session. This is to avoid having to create the db for all the tests. Do you think your new function will be able to be scoped at the session level?

**Comment by zangell44** [comment]
> Do you think your new function will be able to be scoped at the session level?

Yeah it should work fine scoped to the session level

**Comment by davzucky** [comment]
> I'm inclined to wait until the utility in `prefect` is released to avoid needing to remove this code in a later PR. @davzucky does waiting for the next release of Prefect 2.0 block you at all?

Sure, I will keep an eye on the next Orion release. I just manually change the conftest.py manually at the moment.

**Comment by davzucky** [comment]
@zangell44 @desertaxle I updated the fixture to use the new ```prefect@2.0b3``` function ```prefect_test_harness``` as you told me.
Can you review it and check it in? Thank you

**Comment by davzucky** [comment]
Fixed two bugs that I found as well one been related to the latest change:
- SecretManager date comparison should use UTC date. Look like it error at the change of date
- The new ```prefect_test_harness``` function require a new dependency for python 3.7 ```asyncmock; python_version < '3.8'
`` However in Orion you are including ```mock; python_version < '3.8'``` do you know which one is the right one ?`

File: requirements-dev.txt [hunk 3]
```diff
@@ -1,3 +1,4 @@
+asyncmock; python_version < '3.8'
```
Comment (zangell44) [inline requirements-dev.txt:1]:
```suggestion
mock; python_version < '3.8'
```

I think `mock` is what you want here. Do tests still pass?

**Review by zangell44 — APPROVED** [review]
One comment on mock vs. asyncmock, lgtm in general!

**Comment by davzucky** [comment]
@zangell44 Thank you for your comment, indeed I wasn't sure. I found online and tested with ```asyncmock``` and after I committed I check prefect and found that you are using ```mock```.
Updated the code

**Review by desertaxle — APPROVED** [review]
LGTM!

**Pull request merged**